packages/aws_common/lib/src/http/aws_http_client_io.dart

```diff
@@ -0,0 +1,500 @@
+        .tap(onDone: requestProgress.close, (chunk) {
+          requestBytesRead += chunk.length;
+          requestProgress.add(requestBytesRead);
+        })
+        .takeUntil(cancelTrigger)
+        .pipe(ioRequest) as HttpClientResponse;
+
+    if (completer.isCanceled) return;
+    final bodyController = StreamController<List<int>>(sync: true);
+    onCancel = () {
+      logger.verbose('Detaching socket');
+      if (!bodyController.isClosed) {
+        bodyController
+          ..addError(const CancellationException())
+          ..close();
+      }
+      response.detachSocket().then((socket) {
+        socket.destroy();
+      });
+    };
+    response.listen(
+      bodyController.add,
+      onError: bodyController.addError,
+      onDone: () {
+        // Allow `onCancel` to fire if this is the result of a cancellation
+        // before closing body controller.
+        Future.delayed(Duration.zero, bodyController.close);
+      },
+      cancelOnError: true,
+    );
+
+    final headers = <String, String>{};
+    response.headers.forEach((key, values) {
+      headers[key] = values.join(', ');
+    });
+
+    var responseBytesRead = 0;
+    final streamedResponse = AWSStreamedHttpResponse(
+      statusCode: response.statusCode,
+      headers: headers,
+      body: bodyController.stream.tap(
+        (chunk) {
+          responseBytesRead += chunk.length;
+          responseProgress.add(responseBytesRead);
+        },
+        onDone: () {
+          onCancel = null;
+          responseProgress.close();
+        },
+      ),
+    );
+    if (response.headers.contentLength >= 0 &&
+        !response.headers.chunkedTransferEncoding) {
+      completer.complete(streamedResponse.read());
+    } else {
+      completer.complete(streamedResponse);
+    }
+  }
+
+  // Copied from `dart:io`.
+  static bool _isSubdomain(Uri subdomain, Uri domain) {
+    return (subdomain.isScheme(domain.scheme) &&
+        subdomain.port == domain.port &&
+        (subdomain.host == domain.host ||
+            subdomain.host.endsWith('.${domain.host}')));
+  }
+
+  // Copied from `dart:io`.
+  static bool _shouldCopyHeaderOnRedirect(
+    String headerKey,
+    Uri originalUrl,
+    Uri redirectUri,
+  ) {
+    if (_isSubdomain(redirectUri, originalUrl)) {
+      return true;
+    }
+
+    const nonRedirectHeaders = [
+      'authorization',
+      'www-authenticate',
+      'cookie',
+      'cookie2'
+    ];
+    return !nonRedirectHeaders.contains(headerKey.toLowerCase());
+  }
+
+  /// Sends an HTTP/2 request using `package:http2`.
+  Future<void> _sendH2({
+    required AWSBaseHttpRequest request,
+    required AWSLogger logger,
+    required CancelableCompleter<AWSBaseHttpResponse> completer,
+    required Future<void> cancelTrigger,
+    required StreamSink<int> requestProgress,
+    required StreamSink<int> responseProgress,
+  }) async {
+    void Function()? onCancel;
+    unawaited(cancelTrigger.then((_) => onCancel?.call()));
```

Comment:
nit/question: Why is the logging not in this callback same as in _sendH1? I see it below but seems a little inconsistent setup.

Reply:
Updated to be the same